consensus/hotstuff/signature/weighted_signature_aggregator_test.go

```diff
@@ -238,4 +238,30 @@ func TestWeightedSignatureAggregator(t *testing.T) {
 		assert.Nil(t, signers)
 	})
 
+	t.Run("identity aggregated signature", func(t *testing.T) {
+		aggregator, ids, pks, sigs, _, _ := createAggregationData(t, 2)
+		// public key at index 1 is opposite of public key at index 0
+		oppositePk := pks[0].Encode()
+		oppositePk[0] ^= 0x20 // flip the sign bit to flip the point sign
+		var err error
+		pks[1], err = crypto.DecodePublicKey(crypto.BLSBLS12381, oppositePk)
+		require.NoError(t, err)
+
+		// first, add a valid signature
+		_, err = aggregator.TrustedAdd(ids[0].NodeID, sigs[0])
+		require.NoError(t, err)
+
+		// add invalid signature for signer with index 1
+		// (invalid because the corresponding key was altered)
```

Comment:
Thanks for pointing this out. I agree about having tests with minimal causes of errors.
I also think it makes sense in some subtle cases to test the black box implementation returns the expected sentinel error when 2 causes for errors overlap ("expected" means defined by the application). 

In our context, notice that we have updated the logic of building a QC to not rely on the identity signature, but to rely on the identity public key instead. This allows the higher logic to isolate the case of a colluding committee (when signature is identity, we're not sure whether some signatures are invalid or if all signatures are valid and key is identity).
I think the test cases (1) and (2) above do not reflect the expected behaviour because:
    - there should be case (3) for completeness where: the resulting aggregate is the identity signature and some signatures are invalid
    - in cases (1) and (3) there is an overlap that we want to distinguish. Whether the aggregated key is identity or not results in a different sentinel error.


I rather see the unhappy path test cases from the application perspective (for a black box implementation) as:

 I. aggregated key is not identity but some individual signatures are invalid (minimal error) : `model.InvalidSignatureIncludedError`  
  (already [implemented](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/consensus/hotstuff/signature/weighted_signature_aggregator_test.go#L194))
 II. aggregated key is identity and all individual signatures are valid (minimal error) : `model.InvalidAggregatedSignatureError` (missing test!)
 III. aggregated key is identity and some individual signatures are invalid (overlapping errors): `model.InvalidAggregatedSignatureError` ([implemented](https://github.com/onflow/flow-go/blob/35738f2618fb9dd87060955fe5bef0b232341992/consensus/hotstuff/signature/weighted_signature_aggregator_test.go#L241) but the test name was not updated from the previous iteration and hence the confusion!)
 
 For each case (I) (II) and (III), we could add sub tests for (a) identity and (b) non-identity aggregated signatures for better robustness, but that's not relevant for the application logic at this point.

In the above, "invalid signatures" mean they are correctly serialized but do not verify against the public key. 
However I noticed a subtle case when at least a signature does not deserialize correctly:
  - if aggregated key is non-identity, expect `model.InvalidSignatureIncludedError` (minimal error) (already implemented)
  - if aggregated key is identity, expect `model.InvalidSignatureIncludedError` (overlapping error). Why not `model.InvalidAggregatedSignatureError` like the overlapping case above? I think here we didn't reach the cryptographic validity check yet, and we are still checking the collected data prior to the aggregation work. So it makes sense to not speak about an invalid aggregated signature. Happy to hear your thought. 
  
I have implemented the changes in https://github.com/onflow/flow-go/pull/3441/commits/31532375009b84b3cb2afec59a9579b24939079a

Reply:
Fyi, I have updated your suggestion to make `createAggregationData` not instantiate the aggregator (though returning `sks` isn't required here).

Reply:
I've updated the expected returned errors of `Aggregate()` docs in https://github.com/onflow/flow-go/pull/3441/commits/22a585458a9926a7eb18c6de30738211ffbf8c99

Reply:
> if aggregated key is identity, expect `model.InvalidSignatureIncludedError` (overlapping error). Why not `model.IsInvalidAggregatedKeyError` like the overlapping case above? I think here we didn't reach the cryptographic validity check yet, and we are still checking the collected data prior to the aggregation work. So it makes sense to not speak about an invalid aggregated key. Happy to hear your thought.

This is a relatively broad topic. Imagine a function that can return 10 different sentinels. How do we expect the function to react to an input that violates all 10 correctness conditions. Documenting precedence of failure conditions is generally not tractable in that it requires a lot more work and often implementation details need to be exposed to explain the mechanics. 
I like to treat sentinel errors as non-deterministic, where reporting one violation of happy path is sufficient to reject the input. We do no need to (and often can't) return the exhaustive set of _all_ contract violations in the input. 

So in general, I would say the method can just non-deterministically return any one applicable sentinel.

In this particular case, you are making the argument that the more detailed differentiation is useful for the use-case (?) If it is not too much more work, by all means, with a clean documentation, we can explain the precedence of the different failure cases.

I am just saying, we don't have to ... and there does _not_ have to be any particular logic as to why sentinel `errA` is returns over `errB`.

Reply:
thank you, current implementation and test look great.